File: frontend/src/api/push.ts
import { authorizedRequest } from './authorizedRequest'

interface VapidPublicKeyResponse {
  public_key: string
}

export async function fetchVapidPublicKey(): Promise<string> {
  const { public_key: publicKey } = await authorizedRequest<VapidPublicKeyResponse>(
    '/v1/push/vapid-public-key',
  )
  return publicKey
}

export interface PushSubscriptionPayload {
  endpoint: string
  keys: { p256dh: string; auth: string }
}

export function subscribeToPush(payload: PushSubscriptionPayload): Promise<void> {
  return authorizedRequest<void>('/v1/push/subscribe', { method: 'POST', body: payload })
}

export function unsubscribeFromPush(endpoint: string): Promise<void> {
  return authorizedRequest<void>('/v1/push/unsubscribe', {
    method: 'POST',
    body: { endpoint },
  })
}

File: backend/app/api/v1/push.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from app.core.config import settings
from app.db.session import get_db
from app.models.user import User
from app.schemas.push import (
    PushSubscribeRequest,
    PushUnsubscribeRequest,
    VapidPublicKeyResponse,
)
from app.services import push_service

router = APIRouter(prefix="/push", tags=["push"])


@router.get("/vapid-public-key", response_model=VapidPublicKeyResponse)
async def read_vapid_public_key() -> VapidPublicKeyResponse:
    return VapidPublicKeyResponse(public_key=settings.vapid_public_key)


@router.post("/subscribe", status_code=status.HTTP_201_CREATED)
async def subscribe(
    payload: PushSubscribeRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> None:
    await push_service.upsert_subscription(db, current_user.id, payload)


@router.post("/unsubscribe", status_code=status.HTTP_204_NO_CONTENT)
async def unsubscribe(
    payload: PushUnsubscribeRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> None:
    await push_service.remove_subscription(db, current_user.id, payload.endpoint)
